### PR TITLE
Concurrent training processes for project

### PR DESCRIPTION
**Pull request by dcalvom**
**Proposed changes**:

In our special case we need to train multiple models for the same project. This models are independent one from another, but currently Rasa blocks multiple trainings for the same project.

Thus, we added a new global variable for the server that count the number of training processes that the server is running. This variable will increment each time a new training process is launched but will block a new training processes if the `max_training_processes` value passed by parameter to the server has been reached.

In order to keep track of how many processes a project is running, a variable for the project object has been added too in order to keep track of its training processes.

**Status (please check what you already did)**:
- [X] made PR ready for code review
- [X] added some tests for the functionality
- [X] updated the documentation (Don't apply)
- [X] updated the changelog

### CODE REVIEW

File: rasa_nlu/data_router.py [hunk 1]
```diff
@@ -308,18 +311,38 @@ def start_train_process(self,
         def training_callback(model_path):
             model_dir = os.path.basename(os.path.normpath(model_path))
             self.project_store[project].update(model_dir)
+            self._current_training_processes = \
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: I guess `self._current_training_processes -= 1` should do the same?

File: rasa_nlu/data_router.py [hunk 2]
```diff
@@ -308,18 +311,38 @@ def start_train_process(self,
...
+                self._current_training_processes - 1
+            self.project_store[project].current_training_processes = \
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: same

File: rasa_nlu/data_router.py [hunk 3]
```diff
@@ -308,18 +311,38 @@ def start_train_process(self,
...
+                self.project_store[project].current_training_processes \
+                - 1
+            if self.project_store[project].status == 1 \
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: please use `(` around the if clause and avoid `\`

File: rasa_nlu/data_router.py [hunk 4]
```diff
@@ -276,6 +277,8 @@ def get_status(self):
         # be other trainings run in different processes we don't know about.
 
         return {
+            "max_training_processes": self._training_processes,
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: What is the semantic behind this? E.g. what does it mean to set `max_training_processes ` to `5`?

I think there is either an issue in the way I think about this or the way it is implemented. I though the intention was to allow 5 parallel training processes **per project**. The way it is implemented allows 5 process **in total**.

Comment (wrathagom) [reply rasa_nlu/data_router.py]: The pool is created here, setting it to 5 creates 5 process that will receive training requests in a round robin approach. https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/data_router.py#L108

This is part of @PHLF's original changes. We're not changing that, just removing the constraint that restricted it to be 1 training process per project.

**Comment by PHLF** [comment]
Hi there. Nice work. Just to add my 2 cents to the conversation: I have not been really fond of the idea of having multiple processes available for training on the server, as part of the training process is already running concurrently and just one process easily eats much of the available RAM and CPU. This explains the restriction I set up.
Remember that the HTTP server's main purpose is to handle clients requests: the less time you spend at training, the better is your QoS.
I think it is necessary, to enable what you want to achieve, to extract the training process to a separate module you can run on another server if needed.

That were just my 2 cents ;)

File: rasa_nlu/data_router.py (hunk 1 above)
Comment (dcalvom) [reply rasa_nlu/data_router.py]: Sure

File: rasa_nlu/data_router.py (hunk 2 above)
Comment (dcalvom) [reply rasa_nlu/data_router.py]: 👍

File: rasa_nlu/data_router.py (hunk 3 above)
Comment (dcalvom) [reply rasa_nlu/data_router.py]: 👍

**Comment by tmbo** [comment]
I didn't have time yet to take another look at this sorry, there is some stuff in Rasa Core that we need to fix first before I can get back to this.

**Comment by tmbo** [comment]
Thanks a lot for implementing this. Sorry it took so long to get this merged.

**Pull request merged**